rkt/images.go

```diff
@@ -224,49 +222,101 @@ func (f *fetcher) fetchSingleImage(img string, asc string, discover bool) (strin
 		return "", fmt.Errorf("unable to access %q: %v", img, err)
 	}
 
-	if discover && u.Scheme == "" {
+	switch u.Scheme {
+	// image name
+	case "":
+		return f.fetchImageByName(img, ascFile)
+	case "http", "https", "file", "docker":
+		return f.fetchImageByURL(u, ascFile)
+	}
+
+	return "", fmt.Errorf("rkt only supports image name and http, https, docker or file URLs (%s)", img)
+}
+
+func (f *fetcher) fetchImageByName(img string, ascFile *os.File) (string, error) {
+	// check the store
+	if !f.forceRemote {
+		key, err := getStoreKeyFromApp(f.s, img)
+		if err != nil {
+			switch err.(type) {
+			// ignore the error if it's a store.ACINotFoundError
+			case store.ACINotFoundError:
+			default:
+				return "", err
+			}
+		} else {
+			stderr("rkt: using image from local store for image name %s", img)
+			return key, nil
+		}
+	}
+
+	// do remote fetching
+	if !f.noRemote {
+		// Do image discovery
 		if app := newDiscoveryApp(img); app != nil {
-			var discoveryError error
-			if !f.local {
-				stderr("rkt: searching for app image %s", img)
-				ep, err := discoverApp(app, true)
-				if err != nil {
-					discoveryError = err
-				} else {
-					// No specified version label, mark it as latest
-					if _, ok := app.Labels["version"]; !ok {
-						latest = true
-					}
-					return f.fetchImageFromEndpoints(app.Name.String(), ep, ascFile, latest)
+			stderr("rkt: searching for app image %s", img)
+			ep, err := discoverApp(app, true)
+			if err != nil {
+				return "", fmt.Errorf("discovery failed for %q: %v", img, err)
+			} else {
+				latest := false
+				// No specified version label, mark it as latest
+				if _, ok := app.Labels["version"]; !ok {
+					latest = true
 				}
-			}
-			if discoveryError != nil {
-				stderr("discovery failed for %q: %v. Trying to find image in the store.", img, discoveryError)
-			}
-			if f.local || discoveryError != nil {
-				return f.fetchImageFromStore(img)
+				return f.fetchImageFromEndpoints(app.Name.String(), ep, ascFile, latest)
 			}
 		}
 	}
 
-	switch u.Scheme {
-	case "http", "https", "file":
-	case "docker":
-		dockerURL := common.ParseDockerURL(path.Join(u.Host, u.Path))
-		if dockerURL.Tag == "latest" {
-			latest = true
+	return "", fmt.Errorf("unable to fetch image for image name: %s", img)
```

Comment:
nit: I'd like to move this up, so that we can have the regular routine with less indent level:
```go
if app == nil {
    return "", err
}
stderr("rkt: searching for app image %s", im)
ep, err := discoverApp(app, true)
...
```